_posts/2021-11-18-keras-updates/keras-updates.Rmd

```diff
@@ -0,0 +1,159 @@
+
+[^3]: Provided stateless semantics is possible, in some context. This is anything but a matter of course, with an object-oriented language like Python in the background.
+
+However, due to the way TensorFlow and Keras have developed on the Python side -- referring to the big architectural and semantic changes between versions 1.x and 2.x, first comprehensively characterized on this blog [here](https://blogs.rstudio.com/ai/posts/2019-10-08-tf2-whatchanges/) -- it has become more challenging to provide, to the R user, all of the functionality available on the Python side. In addition, maintaining compatibility with several versions of Python TensorFlow -- something R Keras has always been doing -- by necessity gets more and more challenging, the more wrappers and convenience functions you add.
+
+So this is where we complement the above "make it R-like and natural, where possible" with "make it easy to port from Python, where necessary". With the new low-level functionality, you won't have to wait for R wrappers to make use of Python-defined objects. Instead, Python objects may be sub-classed directly from R; and any additional functionality you'd like to add to the subclass is defined in a Python-like syntax. What this means, concretely, is that translating Python code to R has become a lot easier. We'll catch a glimpse of this in the second of our three highlights.
+
+## New in Keras 2.6/7: Three highlights
+
+Among the many new capabilities added in Keras 2.6 and 2.7, we quickly introduce three of the very most important.
+
+-   *Pre-processing layers* significantly help to streamline the training workflow, integrating data manipulation and data augmentation.
+
+-   The ability to subclass Python objects (already alluded to several times) is the new low-level magic that, at the same time, is available to the `keras` user *and* is powering many user-facing enhancements underneath.
+
+-   Recurrent neural network (RNN) layers gain a new cell-level API.
+
+Of these, the first two definitely deserve some deeper treatment; more detailed posts will follow.
+
+### Pre-processing layers
+
+Before the advent of these dedicated layers, pre-processing used to be done as part of the `tfdatasets` pipeline. You would chain operations as required; maybe, integrating random transformations to be applied while training. Depending on what you wanted to achieve, significant programming effort may have ensued.
+
+This is one aspect where the new capabilities can help. Pre-processing layers exist for several types of data, allowing for the usual "data wrangling", as well as data augmentation and feature engineering (as in, hashing categorical data, or vectorizing text).
```

Comment:
```suggestion
This is one area where the new capabilities can help. Pre-processing layers exist for several types of data, allowing for the usual "data wrangling", as well as data augmentation and feature engineering (as in, hashing categorical data, or vectorizing text).
```

Reply:
oops, yep, i guess i was looking for that